Apps/Src/SendCarCAN.c

```diff
@@ -1,36 +1,187 @@
 /**
  * @copyright Copyright (c) 2018-2023 UT Longhorn Racing Solar
  * @file SendCarCAN.c
- * @brief Implements the SendCarCAN Task
+ * @brief Function implementations for the SendCarCAN application.
  * 
- * Resends the feedback from the motor controller
+ * This contains functions relevant to placing CAN messages in a CarCAN queue and periodically sending
+ * those messages in the SendCarCAN task.
  * 
  */
 
-#include "Tasks.h"
-#include "CANbus.h"
-#include "CAN_Queue.h"
 #include "common.h"
+#include "os_cfg_app.h"
+#include "CANbus.h"
+#include "Minions.h"
+#include "Contactors.h"
+#include "Pedals.h"
+#include "Tasks.h"
+#include "SendCarCAN.h"
+#include "SendTritium.h"
+
+#define IO_STATE_DLY_MS 250u 
+
+// Task_PutIOState
+OS_TCB putIOState_TCB;
+CPU_STK putIOState_Stk[TASK_SEND_CAR_CAN_STACK_SIZE];
+
+//fifo
+#define FIFO_TYPE CANDATA_t
+#define FIFO_SIZE 256
+#define FIFO_NAME SendCarCAN_Q
+#include "fifo.h"
+
+// Motor message
+#define NUM_MOTOR_MSGS 15 // Messages received from the motor controller
+#define MOTOR_MSG_BASE_ADDRESS 0x240
+
+static SendCarCAN_Q_t CANFifo; 
+
+static OS_SEM CarCAN_Sem4;
+static OS_MUTEX CarCAN_Mtx;
+
+// Counter array to reduce the frequency at which we send out messages
+static uint8_t motorMsgCount[NUM_MOTOR_MSGS] = {0};
+static uint8_t motorMsgThreshold[NUM_MOTOR_MSGS] = {5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5};
+
+static void Task_PutIOState(void *p_arg);
 
 /**
- * @brief Sends the feedback from the motor controller to be read by telemetry and 
- * sends pedal, switch, light, and contactor information to be read by telemetry
- * 
- * @param p_arg 
- */
+ * @brief print SendCarCAN_Q put/get for debug purposes
+*/
+void print_SendCarCAN_Q(void) {
+    if(SendCarCAN_Q_is_full(&CANFifo)) {
+        printf("\n\rSendCarCAN_Q is full.");
+    }
+    printf("\n\rSendCarCAN_Q put: %d\n\rSendCarCAN_Q get: %d", CANFifo.put, CANFifo.get);
+}
+
+/**
+ * @brief Wrapper to put new message in the CAN queue
+*/
+void SendCarCAN_Put(CANDATA_t message){
+    OS_ERR err;
+    CPU_TS ticks;
+
+    // Reduce the frequency at which we forward motor messages to CarCAN
+    uint8_t msgIdx = message.ID - MOTOR_MSG_BASE_ADDRESS;
+
+    if (msgIdx >= 0 && msgIdx < NUM_MOTOR_MSGS) {  // Check if the message is from the motor controller
+        if (++motorMsgCount[msgIdx] < motorMsgThreshold[msgIdx]) {
+            return;  
+        } else {
```

Comment:
Don't need an else if you're returning, it can just be

`if (++motorMsgCount[msgIdx] < NUM_MOTOR_MSGS) {  
    return;  
}  
motorMsgCount[msgIdx] = 0;`

Reply:
Ah very true!